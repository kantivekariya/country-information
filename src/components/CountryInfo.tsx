import { Outlet } from "react-router-dom";

const CountryInfo = () => {
  return (
    <>
      <h1>Home</h1>
      <Outlet />
    </>
  );
};

export default CountryInfo;
