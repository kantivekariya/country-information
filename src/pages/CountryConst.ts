export const COUNTRY_TEST_CONST = {
  countryTestId: "countryTestId",
};
