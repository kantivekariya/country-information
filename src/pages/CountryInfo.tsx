import { useCallback, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Outlet } from "react-router-dom";
import CardList from "../components/card-list/CardList";
import DropdownComponent from "../components/dropdown/Dropdown";
import Loading from "../components/Loading/Loading";
import SearchBox from "../components/search-box/SearchBox";
import { fetchCountryInfoList } from "../redux/action/CountryInfoAction";
import { COUNTRY_TEST_CONST } from "./CountryConst";

import "./_country-info.scss";

interface IProps {
  theme: string;
}

const CountryInfo = ({ theme }: IProps) => {
  const dispatch = useDispatch<any>();
  const [regions, setRegions] = useState([]);
  const [region, setRegion] = useState("all");
  const [searchField, setSearchField] = useState("");
  const [showTapToTop, setShowTapToTop] = useState(false);

  const { countryList, isLoading } = useSelector(
    (state: any) => state.countryInfo
  );

  useEffect(() => {
    window.addEventListener("scroll", () => {
      if (window.scrollY > 100) {
        setShowTapToTop(true);
      } else {
        setShowTapToTop(false);
      }
    });
  }, []);

  useEffect(() => {
    dispatch(fetchCountryInfoList());
    if (countryList?.length > 1) {
      getRegionsList();
    }
  }, [dispatch]);

  const getRegionsList = useCallback(() => {
    setRegions(
      countryList
        .map((e: { region: any }) => e.region)
        .filter(
          (e: string, index: any, self: string | any[]) =>
            self.indexOf(e) === index && e !== ""
        )
    );
  }, [countryList]);

  /* to handle region sorting */
  const handleRegion = (event: any) => setRegion(event.target.value);

  /* handle search country by name */
  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>) =>
    setSearchField(e.target.value);

  const filteredRegion = useCallback(() => {
    if (region !== "all")
      return countryList?.filter((country: { region: string }) =>
        country?.region?.toLowerCase()?.includes(region?.toLowerCase())
      );
    else return countryList;
  }, [countryList, region]);

  const filterCountries = filteredRegion()?.filter(
    (country: { name: { official: string } }) =>
      country.name?.official
        ?.toLowerCase()
        ?.includes(searchField?.toLowerCase())
  );

  const scrollUp = () => {
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  return (
    <>
      {isLoading ? (
        <Loading theme={theme} />
      ) : (
        <div
          className="container"
          data-testid={COUNTRY_TEST_CONST.countryTestId}
        >
          <div className="controllers">
            <SearchBox
              placeholder={`Search for a country...`}
              handleChange={handleSearch}
              theme={theme}
            />
            <DropdownComponent
              regions={regions}
              title={region !== "all" ? region : "Filter by region"}
              handleChange={handleRegion}
              theme={theme}
            />
          </div>
          <CardList countries={filterCountries} theme={theme} />
        </div>
      )}
      {showTapToTop && (
        <div className="tap-to-top">
          <button color="#fff" className="p-2 " onClick={scrollUp}>
            <i className="fa fa-arrow-up" aria-hidden="true"></i>
          </button>
        </div>
      )}
      <Outlet />
    </>
  );
};

export default CountryInfo;
