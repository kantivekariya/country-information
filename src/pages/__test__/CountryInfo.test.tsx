import { cleanup, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import * as reactRedux from "react-redux";

import "@testing-library/jest-dom";
import CountryInfo from "../CountryInfo";
import { COUNTRY_TEST_CONST } from "../CountryConst";

afterEach(cleanup);

const loaderProps = {
  theme: "light",
};

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
}));

describe("Country Info", () => {
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
  const dummyDispatch = jest.fn();
  useDispatchMock.mockReturnValue(dummyDispatch);

  const mockedState = {
    countryInfo: {
      countryList: [],
      status: "PENDING",
      isLoading: false,
      error: "",
    },
  };

  beforeEach(() => {
    // @ts-ignore
    reactRedux.useSelector.mockImplementation((callback) => {
      return callback(mockedState);
    });
  });

  it("renders country info", () => {
    render(<CountryInfo {...loaderProps} />, {
      wrapper: MemoryRouter,
    });
    const cardListElement = screen.getByTestId(
      COUNTRY_TEST_CONST.countryTestId
    );
    expect(cardListElement).toBeInTheDocument();
    expect(cardListElement).not.toBeChecked();
  });

  it("should be matches snapshot", () => {
    const { baseElement } = render(<CountryInfo {...loaderProps} />, {
      wrapper: MemoryRouter,
    });
    expect(baseElement).toMatchSnapshot();
  });
});
