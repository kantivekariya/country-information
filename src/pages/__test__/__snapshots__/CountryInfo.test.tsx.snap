// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country Info should be matches snapshot 1`] = `
<body>
  <div>
    <div
      class="container"
      data-testid="countryTestId"
    >
      <div
        class="controllers"
      >
        <div
          class="search-container"
          data-testid="searchBoxId"
        >
          <button
            class="search-icon-btn"
          >
            <i
              class="fas fa-search search-icon light-mode"
            />
          </button>
          <input
            class="search light-mode"
            placeholder="Search for a country..."
            type="search"
          />
        </div>
        <div
          class="dropdown-container"
          data-testid="dropDownTestId"
        >
          <button
            class="dropdown-btn light"
            data-testid="manageClickEventId"
          >
            <span>
              Filter by region
            </span>
            <i
              class="fas fa-chevron-down"
            />
          </button>
          <div
            class="dropdown-menu
				initial
				hide
        
			"
          />
        </div>
      </div>
      <div
        class="cardList-container"
        data-testid="cardListTestId"
      />
    </div>
  </div>
</body>
`;
